opentelemetry-exporter-geneva/geneva-uploader/src/payload_encoder/otlp_encoder.rs

```diff
@@ -0,0 +1,411 @@
+use crate::payload_encoder::bond_encoder::{
+    BondDataType, BondEncodedRow, BondEncodedSchema, BondWriter,
+};
+use crate::payload_encoder::central_blob::{CentralBlob, CentralEventEntry, CentralSchemaEntry};
+use chrono::{TimeZone, Utc};
+use opentelemetry_proto::tonic::common::v1::any_value::Value;
+use opentelemetry_proto::tonic::logs::v1::LogRecord;
+use std::collections::HashMap;
+use std::sync::{Arc, RwLock};
+
+type SchemaCache = Arc<RwLock<HashMap<u64, (BondEncodedSchema, Vec<FieldInfo>)>>>;
+type BatchKey = (u64, String);
+type EncodedRow = (String, u8, Vec<u8>); // (event_name, level, row_buffer)
+type BatchValue = (CentralSchemaEntry, Vec<EncodedRow>);
+type LogBatches = HashMap<BatchKey, BatchValue>;
+
+/// Encoder to write OTLP payload in bond form.
+pub struct OtlpEncoder {
+    // TODO - limit cache size or use LRU eviction, and/or add feature flag for caching
+    schema_cache: SchemaCache,
+}
+
+#[derive(Clone, Debug)]
+struct FieldInfo {
+    name: String,
+    type_id: u8,
+    order_id: u16,
+}
+
+impl OtlpEncoder {
+    pub fn new() -> Self {
+        OtlpEncoder {
+            schema_cache: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    pub fn encode_log_batch<'a, I>(
+        &self,
+        logs: I,
+        metadata: &str,
+    ) -> Vec<(u64, String, Vec<u8>, usize)>
+    where
+        I: Iterator<Item = &'a opentelemetry_proto::tonic::logs::v1::LogRecord>,
+    {
+        use std::collections::HashMap;
+
+        let mut batches: LogBatches = HashMap::new();
+
+        for log_record in logs {
+            // 1. Get schema
+            let field_specs = self.determine_fields(log_record);
+            let schema_id = self.calculate_schema_id(&field_specs);
+            let (schema_entry, field_info) = self.get_or_create_schema(schema_id, field_specs);
+
+            // 2. Encode row
+            let row_buffer = self.write_row_data(log_record, &field_info);
+            let event_name = if log_record.event_name.is_empty() {
+                "Log".to_string()
+            } else {
+                log_record.event_name.clone()
+            };
+            let level = log_record.severity_number as u8;
+
+            // 3. Insert into batches - Key is (schema_id, event_name)
+            batches
```

Comment:
I'm now wondering if it's even worth tracking the different schemas. Could we just track `event_name` and serialize whatever fields the corresponding log records have?

Reply:
Yes, exactly my thoughts. It should be fine to have the events with different schema under the same table name. Will track this as part of #306